feat: add terminal formatter

guilhermeprokisch opened the pull request:
Hi @Colonial-Dev,

Here is a draft for implementing the terminal formatter. My Rust skills are not the best, so there might be some bugs in it. I was able to make it work in one of my projects, so it should be at least on the right track. I'm sharing more to ensure that the work is not lost.

guilhermeprokisch committed:
feat: add terminal formatter

guilhermeprokisch committed:
feat: add terminal formatter

Colonial-Dev commented on src/formatter/terminal_themed.rs:
```diff
@@ -0,0 +1,76 @@
+#[derive(Debug)]
+pub struct TerminalFormatter {
+    theme: Arc<Theme>,
+    stdout: RefCell<StandardStream>,
```
`RefCell<T>` isn't `Send` or `Sync` (that is, multi-thread-compatible) like `Arc<T>` is.

For something like this, you *probably* don't want to be writing to standard output/error from multiple threads anyways, so I would just do something like:

```rs
#[derive(Debug, Clone)]
pub struct Terminal {
    theme: Theme,
    stream: StandardSteam
}
```

Colonial-Dev commented on src/formatter/terminal_themed.rs:
```diff
@@ -0,0 +1,76 @@
+/// This formatter holds theme data in an [`Arc<Theme>`], making it cheap to clone and usable
+/// from multiple threads.
+#[derive(Debug)]
+pub struct TerminalFormatter {
```
I would change this to just be `Terminal`, similar to how the basic HTML formatter is just called `Html`.

Colonial-Dev commented on src/formatter/terminal_themed.rs:
```diff
@@ -0,0 +1,76 @@
+    pub fn new(theme: Theme) -> Self {
+        Self {
+            theme: Arc::new(theme),
+            stdout: RefCell::new(StandardStream::stdout(ColorChoice::Always)),
```
I wouldn't assume that users want standard output - consider something like printing code snippets in a backtrace to standard error. Maybe take a `StandardStream` as one of the parameters.

Colonial-Dev commented on src/formatter/terminal_themed.rs:
```diff
@@ -0,0 +1,76 @@
+                let text = &source[start..end];
+                let mut stdout = self.stdout.borrow_mut();
+                stdout.write_all(text.as_bytes())?;
+                stdout.flush()?;
```
I would implement `finish` and call `flush` in there - doing it for every event could get mighty expensive (unless the immediacy is important to your use case.)

Colonial-Dev reviewed (changes requested):
Pretty good, thank you! I don't see any bugs, just a few things that need tweaking.

guilhermeprokisch committed:
feat: review and examples

guilhermeprokisch commented:
@Colonial-Dev, I have addressed some of your reviews. Please take a look and let me know if it meets your approval now. I have also added an example folder demonstrating how to use it, making it easy to see the results by running `cargo run --example terminal_highlight`. Additionally, I have included a theme generator. Frankly, this was the most challenging part, in my opinion. I believe that shipping with some built-in themes would be better.

guilhermeprokisch committed:
feat: reviews and example

Colonial-Dev committed:
Update theme_example.rs

Add a 'main' function that prints the generated theme, so the example will compile.

Colonial-Dev committed:
TIL examples can depend on each other!

Colonial-Dev commented:
Looks good to me!

> I believe that shipping with some built-in themes would be better.

Yeah, I feel you. The Helix theme support is almost done!

guilhermeprokisch commented:
@Colonial-Dev  Btw I'm building a project based on this feature 

https://github.com/guilhermeprokisch/see/

Anything I'm here to help inkjet maintenance
